HCCDOC-4977: Clean up dynamic plugins overview and fix duplications (#111188)
openshift/openshift-docs: web_console/dynamic-plugin/overview-dynamic-plugin.adoc

Reviewer: Maybe outside of scope, but the id for an assembly should not have a context variable.

Author: this is an inherited doc but good call. can i simply remove it @dfitzmau?

Reviewer: Yes, you can simply remove it. Thanks!

:_mod-docs-content-type: ASSEMBLY
include::_attributes/common-attributes.adoc[]
include::_attributes/attributes-openshift-dedicated.adoc[]
[id="overview-dynamic-plugin"]
= Overview of dynamic plugins

> 

Reviewer 2: This change may or may not matter, but ultimately I'm going to err on the side of caution here--

Please don't make this change. Undo it and add line 6 back in. The Tools team has asked that we don't make any changes to IDs because they're overwhelmed with requests right now. As it stands, I don't _think_ that this would actually affect the URL, but I did talk to Kathryn and we both agree--better safe than sorry.

Reviewer 3: FWIW both Darragh and I told her to fix this

Reviewer 2: Sorry :D (fwiw I really don't think it would matter, but alas)

Reviewer 3: No biggie, just wanted to make sure it was clear she was told to do this because I know that back-and-forth is very frustrating. Sorry, Jocelyn! :bow:

Author: @stevsmit thanks for flagging. Since Jeana confirmed they and Darragh had asked me to make the `:context:`/ID change, I've kept it as-is. no undo needed there.

on the cherry-pick question: Sam (@spadgett) said going back to just 5.0 and 4.22 seems reasonable, so 4.21 might not actually need this. I have the branch/enterprise-4.21 label on here already though. want me to drop it, or would you rather keep it targeted to 4.21 too just to be safe?

also just pushed a fix to the "exposes the application" wording per @logonoff confirmation (should be "exposes the plugin"). that was the other open thread. 

I think that's everything now. eager for your responses and thanks again.

toc::[]

[role="_abstract"]

> The OpenShift web console loads and interprets dynamic plugins from remote sources at runtime. One way to deliver and expose dynamic plugins to the console is through OLM Operators. The Operator creates a deployment on the platform with an HTTP server to host the plugin and exposes it using a Kubernetes service.

Reviewer: ```suggestion
The {product-title} web console loads and interprets dynamic plugins from remote sources at runtime. One way to deliver and expose dynamic plugins to the console is through OLM Operators. The Operator creates a deployment on the platform with an HTTP server to host the plugin and exposes the plugin using a Kubernetes service.
```

exposes the application or the plugin?

Author: @logonoff or @spadgett can you verify which it is? good eye, @dfitzmau.

Reviewer 4: I think the plugin. console operator exposes the host application

[id="general-plug-in-guidelines"]
== General guidelines
Follow these guidelines when you create a plugin:

> * You must use link:https://nodejs.org/en/[`Node.js`] to build and run your plugin. You can use any package manager, such as npm or yarn.

Author: n/a. assembly

[id="general-plug-in-guidelines"]
== General guidelines
When creating your plugin, follow these general guidelines:
Follow these guidelines when you create a plugin:

* link:https://nodejs.org/en/[`Node.js`] and link:https://yarnpkg.com/[`yarn`] are required to build and run your plugin.
* You must use link:https://nodejs.org/en/[`Node.js`] and link:https://yarnpkg.com/[`yarn`] to build and run your plugin.
* Prefix your CSS class names with your plugin name to avoid collisions. For example, `my-plugin_\_heading` and `my-plugin_\_icon`.
* Maintain a consistent look, feel, and behavior with other console pages.
* Follow link:https://www.i18next.com/[react-i18next] localization guidelines when creating your plugin. You can use the `useTranslation` hook like the one in the following example:
+
[source,tsx,subs="+quotes,+macros"]
----
conster Header: React.FC ++= () => {++
const Header: React.FC ++= () => {++
  ++const { t } = useTranslation('plugin__console-demo-plugin');++
  ++return <h1>{t('Hello, World!')}</h1>;++
++};++
----

> * Avoid selectors that could affect markup outside of your plugins components, such as element selectors. These are not APIs and are subject to change. Using them might break your plugin. Avoid selectors like element selectors that could affect markup outside of your plugins components.

Reviewer 5: No need to remove this: `Avoid selectors like element selectors that could affect markup outside of your plugins components.`

[id="general-plug-in-guidelines"]
== General guidelines
When creating your plugin, follow these general guidelines:
Follow these guidelines when you create a plugin:

> * link:https://nodejs.org/en/[`Node.js`] and link:https://yarnpkg.com/[`yarn`] are required to build and run your plugin.

Reviewer 4: Yarn is not required but you do need node.js (many plugins switched to npm nowadays so you're free to use any package manager)

> * You must use link:https://nodejs.org/en/[`Node.js`] and link:https://yarnpkg.com/[`yarn`] to build and run your plugin.

Author: n/a this is an assembly

> * You must use link:https://nodejs.org/en/[`Node.js`] and link:https://yarnpkg.com/[`yarn`] to build and run your plugin.

Author: n/a this is an assembly

* Prefix your CSS class names with your plugin name to avoid collisions. For example, `my-plugin_\_heading` and `my-plugin_\_icon`.
* Maintain a consistent look, feel, and behavior with other console pages.
* Follow link:https://www.i18next.com/[react-i18next] localization guidelines when creating your plugin. You can use the `useTranslation` hook like the one in the following example:
+
[source,tsx,subs="+quotes,+macros"]
----
conster Header: React.FC ++= () => {++
const Header: React.FC ++= () => {++
  ++const { t } = useTranslation('plugin__console-demo-plugin');++
  ++return <h1>{t('Hello, World!')}</h1>;++
++};++
----

* Avoid selectors that could affect markup outside of your plugins components, such as element selectors. These are not APIs and are subject to change. Using them might break your plugin. Avoid selectors like element selectors that could affect markup outside of your plugins components.
* Provide valid JavaScript Multipurpose Internet Mail Extension (MIME) type using the `Content-Type` response header for all assets served by your plugin web server. Each plugin deployment should include a web server that hosts the generated assets of the given plugin.

> * You must build your plugin with Webpack using Webpack version 5 and later.

Reviewer 6: Looks like Webapack is no more mentioned? Is that correct? I don't know.

Author: @logonoff can you confirm?

Reviewer 4: Webpack is still required for now but the console plugin template should guide users towards the suggested setup already

Author: @logonoff I rewrote it like this: * You must build your plugin with Webpack version 5 or later. Refer to the console plugin template for the suggested setup.

ok?

also, where can users find the console plugin template?

Reviewer 4: LGTM

https://github.com/openshift/console-plugin-template/

> * You should prefix CSS class names with your plugin name to avoid collisions. For example, `my-plugin_\_heading` and `my-plugin_\_icon`.

Reviewer 4: Prefixing CSS classes is important and this should be kept in the docs

> * Use link:https://www.patternfly.org/components/all-components/[PatternFly] components and PatternFly CSS variables. The SDK provides core PatternFly components. Using PatternFly components and variables help your plugin look consistent in future console versions.

Author: n/a this is an assembly

> * Use link:https://www.patternfly.org/components/all-components/[PatternFly] components and PatternFly CSS variables. The SDK provides core PatternFly components. Using PatternFly components and variables help your plugin look consistent in future console versions.

Reviewer 4: The SDK does not provide patternfly components, you are expected to just consume patternfly components directly. The SDK does provide console-specific wrappers around certain patternfly components however (e.g., SDK `ListPageHeader` wraps around PatternFly `PageHeader`)

[id="general-plug-in-guidelines"]
== General guidelines
Follow these guidelines when you create a plugin:

> * You must use link:https://nodejs.org/en/[`Node.js`] to build and run your plugin. You can use any package manager, including npm, yarn, and so on.

Author: n/a this is an assembly

* Follow link:https://www.i18next.com/[react-i18next] localization guidelines when creating your plugin. You can use the `useTranslation` hook like the one in the following example:
+
[source,tsx,subs="+quotes,+macros"]
----
const Header: React.FC ++= () => {++
  ++const { t } = useTranslation('plugin__console-demo-plugin');++
  ++return <h1>{t('Hello, World!')}</h1>;++
++};++
----

* Avoid selectors that could affect markup outside of your plugin components, such as element selectors. These are not APIs and are subject to change. Using them might break your plugin. To prevent this, do not use selectors that change HTML elements outside of your own plugin.

* Provide valid JavaScript Multipurpose Internet Mail Extension (MIME) type using the `Content-Type` response header for all assets your plugin web server hosts. Each plugin deployment should include a web server that hosts the plugin's generated assets.

* Prefix CSS class names with your plugin name to avoid collisions; for example, `my-plugin_\_heading` and `my-plugin_\_icon`.

== PatternFly guidelines
When creating your plugin, follow these guidelines for using PatternFly:

> * Use link:https://www.patternfly.org/components/all-components/[PatternFly] components and CSS variables directly. The SDK provides console-specific wrappers around certain PatternFly components; for example, the SDK's `ListPageHeader` wraps PatternFly's `PageHeader`. Using PatternFly components and variables helps your plugin look consistent in future console versions.

Author: assembly

[id="general-plug-in-guidelines"]
== General guidelines
Follow these guidelines when you create a plugin:

> * You must use link:https://nodejs.org/en/[`Node.js`] to build and run your plugin. You can use any package manager, including npm, yarn, and so on.

Reviewer: ```suggestion
* You must use link:https://nodejs.org/en/[`Node.js`] to build and run your plugin. You can use any package manager, such as npm or yarn.
```

* Prefix your CSS class names with your plugin name to avoid collisions. For example, `my-plugin_\_heading` and `my-plugin_\_icon`.
* Maintain a consistent look, feel, and behavior with other console pages.
* Follow link:https://www.i18next.com/[react-i18next] localization guidelines when creating your plugin. You can use the `useTranslation` hook like the one in the following example:
+
[source,tsx,subs="+quotes,+macros"]
----
const Header: React.FC ++= () => {++
  ++const { t } = useTranslation('plugin__console-demo-plugin');++
  ++return <h1>{t('Hello, World!')}</h1>;++
++};++
----

* Avoid selectors that could affect markup outside of your plugin components, such as element selectors. These are not APIs and are subject to change. Using them might break your plugin. To prevent this, do not use selectors that change HTML elements outside of your own plugin.

> * Provide valid JavaScript Multipurpose Internet Mail Extension (MIME) type using the `Content-Type` response header for all assets your plugin web server hosts. Each plugin deployment should include a web server that hosts the plugin's generated assets.

Reviewer: ```suggestion
* Provide valid JavaScript Multipurpose Internet Mail Extension (MIME) type by using the `Content-Type` response header for all assets your plugin web server hosts. Each plugin deployment should include a web server that hosts the generated assets of the plugin.
```
To avoid giving possessive qualities to inanimate objects.

> * Provide valid JavaScript Multipurpose Internet Mail Extension (MIME) type using the `Content-Type` response header for all assets your plugin web server hosts. Each plugin deployment should include a web server that hosts the plugin's generated assets.

Reviewer: Should or must?

Author: another good question. @logonoff or @spadgett?

Reviewer 4: Each plugin deployment must include a web server that hosts the plugin's generated assets

> * You must build your plugin with Webpack version 5 or later. Refer to the link:https://github.com/openshift/console-plugin-template/[console plugin template] for the suggested setup.

Author: n/a assembly
